central/apitoken/expiration/expiration.go

```diff
@@ -0,0 +1,186 @@
+	staleNotificationTimestamp := protoconv.ConvertTimeToTimestamp(staleNotificationDate)
+
+	scheduleCtx := sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Notifications),
+		),
+	)
+	var notificationSchedule *storage.NotificationSchedule
+	notificationSchedule, found, err := n.store.GetNotificationSchedule(scheduleCtx)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	if !found || notificationSchedule == nil {
+		notificationSchedule = &storage.NotificationSchedule{
+			LastRun: protoconv.ConvertTimeToTimestamp(staleNotificationDate.Add(-1 * time.Hour)),
+		}
+	}
+	if notificationSchedule.GetLastRun().Compare(staleNotificationTimestamp) >= 0 {
+		return
+	}
+	notificationSchedule.LastRun = protoconv.ConvertTimeToTimestamp(now)
+	err = n.store.UpsertNotificationSchedule(scheduleCtx, notificationSchedule)
+	if err != nil {
+		log.Error(errors.Wrap(err, "failed to update expired API token notification last run"))
+	}
+
+	expiringTokens, err := n.listItemsToNotify(now, aboutToExpireDate)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	err = n.notify(expiringTokens)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+}
+
+func (n *expirationNotifierImpl) listItemsToNotify(now time.Time, expiresUntil time.Time) ([]*storage.TokenMetadata, error) {
+	formattedNow := now.Format(timestampLayout)
+	formattedExpiresUntil := expiresUntil.Format(timestampLayout)
+	// Search tokens that expire before expiresUntil, that have not expired yet,
+	// and that have not been notified since notifiedUntil.
+	// That is Expiration < expiresUntil and LastNotified < notifiedUntil.
+	queryNotExpired := search.NewQueryBuilder().
+		AddStrings(search.Expiration, fmt.Sprintf(">%s", formattedNow)).
+		ProtoQuery()
+	queryAboutToExpire := search.NewQueryBuilder().
+		AddStrings(search.Expiration, fmt.Sprintf("<%s", formattedExpiresUntil)).
+		ProtoQuery()
+	queryNotRevoked := search.NewQueryBuilder().
+		AddBools(search.Revoked, false).
+		ProtoQuery()
+	query := search.ConjunctionQuery(
+		queryNotExpired,
+		queryAboutToExpire,
+		queryNotRevoked,
+	)
+	response, err := n.store.SearchRawTokens(expirySearchCtx, query)
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func (n *expirationNotifierImpl) notify(items []*storage.TokenMetadata) error {
+	now := time.Now()
+	for _, token := range items {
+		expiration := protoconv.ConvertTimestampToTimeOrNow(token.GetExpiration())
+		ttl := expiration.Sub(now)
+		timeUnit := int(expirationSlice.Seconds())
+		ttlSeconds := int(ttl.Seconds())
+		sliceCount := ttlSeconds / timeUnit
+		if ttlSeconds%timeUnit != 0 {
+			sliceCount++
+		}
+		sliceDuration := "hours"
+		if sliceCount == 1 {
+			sliceDuration = "hour"
+		}
+		log.Warnf("API Token %s (ID %s) will expire in less than %d %s.", token.GetName(), token.GetId(), sliceCount, sliceDuration)
```

Comment:
So, are there any plans to add other notification methods as well?

Right now, this is quite simplistic only for the API token one, but as mentioned within the initial design doc it might not be sufficient enough for ACSCS users, as they have limited access to the logs.

In light of that, it might be also good to abstract the `notify` mechanism more than just this simple method, and provide an interface `Notifier`  which provides:
```go
type Notifier interface {
  Notify(token *storage.TokenMetadata) error
}
```
and create a "log notifier" implementation here. This way when we decide to add more "notifiers" (e.g. email, audit logs) the code here doesn't need to be completely refactored.

Reply:
This depends on the inputs that come from the product side.
I do agree that in the longer run, there could be multiple notification channels, and that an interface for notification would then make sense. I would rather consider the bulk of tokens about to expire rather than individual tokens as input for a notifier (e.g. I'd rather send a single e-mail containing all tokens about to expire rather than an email for each token about to expire).